File: madhav/doc_events/production_plan.py
import frappe
from frappe import _
import traceback


def duplicate_po_items_to_assembly_items_without_consolidate(doc, method):

    if not getattr(doc, "po_items", None):
        return

    fields_to_copy = [
        "item_code",
        "section_weight",
        "bom_no",
        "planned_qty",
        "pending_qty",
        "pieces",
        "length",
        "length_size_m",
        "stock_uom",
        "warehouse",
        "planned_start_date",
        "product_bundle_item",
        "sales_order",
        "sales_order_item",
        "description",
        "customer",
        "customer_name",
        "customers_purchase_order",
    ]

    # Build index of existing target rows
    existing_map = {
        (row.item_code, row.sales_order_item): row
        for row in doc.get("assembly_items_without_consolidate", [])
    }

    for row in doc.get("po_items"):
        key = (row.item_code, row.sales_order_item)

        if key in existing_map:
            # Update existing row → no reset
            target_row = existing_map[key]
        else:
            # Append new row
            target_row = doc.append("assembly_items_without_consolidate", {})

        for field in fields_to_copy:
            target_row.set(field, getattr(row, field, None))


def consolidate_assembly_items(doc, method):

    if not doc.po_items:
        return

    consolidated = {}

    for row in doc.po_items:
        key = row.item_code

        if key not in consolidated:
            consolidated[key] = row
        else:
            base = consolidated[key]

            base.planned_qty += row.planned_qty or 0
            base.pieces += row.pieces or 0
            base.length += row.length or 0
            base.length_size_m += row.length_size_m or 0

            # Remove duplicate rows safely
            doc.remove(row)

    # Only update calculated fields, never delete rows
    for row in consolidated.values():
        row.section_weight = frappe.db.get_value(
            "Item", row.item_code, "weight_per_meter"
        )


def update_so_pieces(doc, method=None):
    aggregated = {}
    for item in doc.po_items:
        if not item.sales_order_item:
            continue
        if item.sales_order_item not in aggregated:
            aggregated[item.sales_order_item] = {"pieces": 0}
        aggregated[item.sales_order_item]["pieces"] += item.pieces or 0

    for so_item_name, totals in aggregated.items():
        so_item = frappe.db.get_value(
            "Sales Order Item",
            so_item_name,
            ["pieces", "production_plan_pieces"],
            as_dict=True
        )
        if not so_item:
            continue

        frappe.db.set_value(
            "Sales Order Item",
            so_item_name,
            {
                "pieces":                 (so_item.pieces or 0)                 - totals["pieces"],
                "production_plan_pieces": (so_item.production_plan_pieces or 0) + totals["pieces"],
            },
            update_modified=False
        )


def revert_so_pieces(doc, method=None):
    aggregated = {}
    for item in doc.po_items:
        if not item.sales_order_item:
            continue
        if item.sales_order_item not in aggregated:
            aggregated[item.sales_order_item] = {"pieces": 0}
        aggregated[item.sales_order_item]["pieces"] += item.pieces or 0

    for so_item_name, totals in aggregated.items():
        so_item = frappe.db.get_value(
            "Sales Order Item",
            so_item_name,
            ["pieces", "production_plan_pieces"],
            as_dict=True
        )
        if not so_item:
            continue

        frappe.db.set_value(
            "Sales Order Item",
            so_item_name,
            {
                "pieces":                 (so_item.pieces or 0)                 + totals["pieces"],
                "production_plan_pieces": (so_item.production_plan_pieces or 0) - totals["pieces"],
            },
            update_modified=False
        )

File: madhav/public/js/production_plan.js
frappe.ui.form.on('Production Plan', {
    refresh: function(frm) {
        if (frm.doc.docstatus === 1) {
            if (frm.doc.status !== "Completed") {
                let items = frm.events.get_items_for_work_order(frm);

				if (items?.length && frm.doc.status !== "Closed") {
                    frm.remove_custom_button('Work Order / Subcontract PO', 'Create');
                    frm.add_custom_button(
                        __("Work Order or Subcontract PO"),
                        () => {
                            frm.trigger("make_work_order_custom");
                            // console.log("triggered")
                        },
                        __("Create")
                    );
                }
            }
        }
        // Show button only if Production Plan is submitted
        if (frm.doc.docstatus === 1) {
            frm.add_custom_button('Create Cutting Plan', function() {
                let source_warehouse = frm.doc.source_warehouse;
                let target_warehouse = frm.doc.fg_warehouse;

                frappe.new_doc("Cutting Plan", {
                    production_plan: frm.doc.name,  // link back to Production Plan
                    company: frm.doc.company,
                    target_qty: frm.doc.total_planned_qty,  // use Production Plan qty field
                    default_source_warehouse: source_warehouse,
                    default_finished_goods_warehouse: target_warehouse,
                    date: frappe.datetime.now_datetime(),
                });

                // Wait for Cutting Plan form to open
                setTimeout(function() {
                    let cutting_plan_form = cur_frm;
                    if (cutting_plan_form && cutting_plan_form.doctype === 'Cutting Plan') {
                        cutting_plan_form.set_value('default_source_warehouse', source_warehouse);

                        // Map items from Production Plan child table
                        let production_plan_items = frm.doc.po_items || [];

                        production_plan_items.forEach(d => {
                            let row = cutting_plan_form.add_child('cut_plan_detail');
                            row.item_code = d.item_code;
                            row.source_warehouse = d.warehouse || source_warehouse;
                            row.qty = d.planned_qty || d.qty;
                        });

                        cutting_plan_form.refresh_field('cut_plan_detail');
                    }
                }, 1000);
            });
        }
    },
    get_items_for_work_order(frm) {
		let items = frm.doc.po_items;
		if (frm.doc.sub_assembly_items?.length) {
			items = [...items, ...frm.doc.sub_assembly_items];
		}

		let has_items =
			items.filter((item) => {
				if (item.planned_qty) {
					return item.planned_qty > item.ordered_qty;
				} else {
					return item.qty > (item.received_qty || item.ordered_qty);
				}
			}) || [];

		return has_items;
	},
    make_work_order_custom(frm) {
		frappe.call({
			method: "make_work_order",
			freeze: true,
			doc: frm.doc,
			callback: function () {
				frm.reload_doc();
                update_latest_work_order(frm);
			},
		});
	},
    get_items(frm) {
        // Wait for ERPNext to finish populating rows
        frappe.after_ajax(() => {
            populate_all_rows(frm);
        });
    },
    get_sales_orders(frm) {
        // Delay required because rows are added asynchronously
        setTimeout(() => {
            populate_customer_names(frm);
        }, 500);
    },
});
function update_latest_work_order(frm) {
    frappe.call({
        method: "madhav.api.update_latest_wo_from_pp", // server method
        args: {
            production_plan: frm.doc.name
        },
        freeze: true,
        callback: function () {
            frm.reload_doc();
            frappe.msgprint("✅ Work Order updated from Production Plan!");
        }
    });
}
function populate_customer_names(frm) {
    (frm.doc.sales_orders || []).forEach(row => {
        if (row.sales_order && !row.customer_name) {
            frappe.db.get_value(
                "Sales Order",
                row.sales_order,
                ["customer", "customer_name"],
                (r) => {
                    if (r) {
                        row.customer = r.customer;
                        row.customer_name = r.customer_name;
                        frm.refresh_field("sales_orders");
                    }
                }
            );
        }
    });
}

frappe.ui.form.on("Production Plan Item", {
    item_code: function (frm, cdt, cdn) {
        fetch_so_item_pcs(frm, cdt, cdn);
    },
});
function populate_all_rows(frm) {
    (frm.doc.po_items || []).forEach(row => {
        fetch_so_item_pcs(frm, row.doctype, row.name);
    });
}
function fetch_so_item_pcs(frm, cdt, cdn) {
    let row = locals[cdt][cdn];

    if (!row.sales_order || !row.item_code) return;

    frappe.call({
        method: "madhav.api.get_so_item_pcs",
        args: {
            sales_order: row.sales_order,
            item_code: row.item_code,
            sales_order_item: row.sales_order_item,
            planned_qty: row.planned_qty || 0,
            row_id: cdn
        },
        callback(r) {
            console.log(r)
            if (r.message) {
                const row_id = r.message.row_id;
    
                frappe.model.set_value(cdt, row_id, "length_size_m", r.message.length_size || 0);
                // frappe.model.set_value(cdt, row_id, "pieces", r.message.pieces || 0);
                frappe.model.set_value(cdt, row_id, "po_no", r.message.po_no || "");
                // frappe.model.set_value(cdt, row_id, "section_weight", r.message.total_weight || "");
                frappe.model.set_value(cdt, row_id, "planned_qty", r.message.planned_qty || 0);
                frappe.model.set_value(cdt, row_id, "assorted_length", r.message.assorted_length || "");
                frappe.model.set_value(cdt, row_id, "pending_qty", r.message.planned_qty || 0);
                frappe.model.set_value(cdt, row_id, "customers_purchase_order", r.message.po_no || "");
                frappe.model.set_value(cdt, row_id, "customer", r.message.customer || "");
                frappe.model.set_value(cdt, row_id, "customer_name", r.message.customer_name || "");
            }
        }
    });
}

frappe.ui.form.on("Production Plan Sales Order", {
    sales_order(frm, cdt, cdn) {
        const row = locals[cdt][cdn];

        if (!row.sales_order) {
            row.customer_name = "";
            frm.refresh_field("sales_orders");
            return;
        }

        frappe.db.get_value(
            "Sales Order",
            row.sales_order,
            ["customer", "customer_name"],
            (r) => {
                if (r) {
                    row.customer = r.customer;
                    row.customer_name = r.customer_name;
                    frm.refresh_field("sales_orders");
                }
            }
        );
    }
});
